homeassistant/components/nanoleaf/light.py

```diff
@@ -53,69 +57,40 @@
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+def setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Import Nanoleaf light platform."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data={CONF_HOST: config[CONF_HOST], CONF_TOKEN: config[CONF_TOKEN]},
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, add_entities: AddEntitiesCallback
+) -> None:
     """Set up the Nanoleaf light."""
-
-    if DATA_NANOLEAF not in hass.data:
-        hass.data[DATA_NANOLEAF] = {}
-
-    token = ""
-    if discovery_info is not None:
-        host = discovery_info["host"]
-        name = None
-        device_id = discovery_info["properties"]["id"]
-
-        # if device already exists via config, skip discovery setup
-        if host in hass.data[DATA_NANOLEAF]:
-            return
-        _LOGGER.info("Discovered a new Nanoleaf: %s", discovery_info)
-        conf = load_json(hass.config.path(CONFIG_FILE))
-        if host in conf and device_id not in conf:
-            conf[device_id] = conf.pop(host)
-            save_json(hass.config.path(CONFIG_FILE), conf)
-        token = conf.get(device_id, {}).get("token", "")
-    else:
-        host = config[CONF_HOST]
-        name = config[CONF_NAME]
-        token = config[CONF_TOKEN]
-
-    nanoleaf_light = Nanoleaf(host)
-
-    if not token:
-        token = nanoleaf_light.request_token()
-        if not token:
-            _LOGGER.error(
-                "Could not generate the auth token, did you press "
-                "and hold the power button on %s"
-                "for 5-7 seconds?",
-                name,
-            )
-            return
-        conf = load_json(hass.config.path(CONFIG_FILE))
-        conf[host] = {"token": token}
-        save_json(hass.config.path(CONFIG_FILE), conf)
-
-    nanoleaf_light.token = token
-
-    try:
-        info = nanoleaf_light.info
-    except Unavailable:
-        _LOGGER.error("Could not connect to Nanoleaf Light: %s on %s", name, host)
-        return
-
-    if name is None:
-        name = info.name
-
-    hass.data[DATA_NANOLEAF][host] = nanoleaf_light
-    add_entities([NanoleafLight(nanoleaf_light, name)], True)
+    nanoleaf = Nanoleaf(entry.data[CONF_HOST])
+    nanoleaf.token = entry.data[CONF_TOKEN]
+    info = await hass.async_add_executor_job(pynanoleaf_get_info, nanoleaf)
+    name = info["name"]
+    serial_no = info["serialNo"]
+    add_entities([NanoleafLight(nanoleaf, name, serial_no)], True)
```

Comment:
Why pass in name and unique ID? The entity can just fetch those from the nanoleaf object directly

Reply:
I can't do I/O in  `__init__` and `unique_id`.
> RuntimeError: I/O must be done in the executor; Use `await hass.async_add_executor_job()`

I can't use `await` in a synchronous environment and hass is None in `__init__`??

Reply:
I'm confused. Is fetching the name or unique ID doing I/O ?

You can do it inside `async_added_to_hass`

Reply:
> I'm confused. Is fetching the name or unique ID doing I/O ?

Yes, some properties make requests. That's why I had to create a wrapper function.
 https://github.com/home-assistant/core/blob/21bf79ffdc4d4b2adb2a2125737a3948a56d6d24/homeassistant/components/nanoleaf/util.py#L5-L7

> You can do it inside `async_added_to_hass`

This doesn't work because unique_id and name are retrieved before `async_added_to_hass` is executed.

Reply:
This was confusing for me because the name is the unique id.

https://github.com/home-assistant/core/pull/52199#discussion_r694041810

Suggest adding comments like suggested above.